proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,217 @@
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
+  </packageSources>
+  <auditSources>
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
+  </auditSources>
+</configuration>
+```
+
+Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
+However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.
+
+NuGet will use same semantics for `auditSources` as it already has with `packageSources`.
+This means that `auditSources` are accumulated from other nuget.config files, and `<clear />` can be used to remove them.
+`auditSources` will also follow `packageSources` with regards to insecure http warnings, errors, and `allowInsecureConnections` configurations.
+Sources that need credentials will use `packageSourceCredentials`, using the `key` as the lookup, just as `packageSources` does.
+This means that credentials can be re-used by using the same key for a source that is defined in both `packageSources` and `auditSources`.
+
+#### Restore summary
+
+When restore is run at normal or detailed verbosity, NuGet outputs detected nuget.config files and package sources (feeds) used.
+Audit sources should be added.
+
+```diff
+ NuGet Config files used:
+     C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config
+
+ Feeds used:
+     https://api.nuget.org/v3/index.json
++
++Audit sources used:
++    https://api.nuget.org/v3/index.json
+```
+
+When `auditSources` do not have any configured sources, replay all of the package sources again under the audit sources used section, to make it explicit to customers that these sources are being used.
```

Comment:
I need to check the implementation, but if "feeds used" (`packageSources`) doesn't take into account package sources defind via msbuild, then the output will be incorrect. On the other hand, if only one project uses an additional source via MSBuild, then it might be confusing why the source is listed when most projects don't use it?  I'm really not a fan of NuGet settings via MSBuild, it's a shame it's adding additional complexity/uncertainty.

Also, another idea is have a message saying "Package sources were used as audit sources", and avoid duplicating the list again.

Anyone have feedback?

I'm actually learning towards the "Package sources were used as audit sources" message, I just don't feel like locally editing the markdown again and pushing another commit unless others agree.

What's better?

```text
Audit sources used:
    Package feed(s) used as audit source(s).
```

or just a one line

```text
Package feed(s) used as audit source(s).
```

Reply:
Outside of Microsoft it is not uncommon for users to have 5+ sources. So multi-line is ideal here. I can share more survey results but I think I shared this specific question with you recently if you want to double check.

Reply:
in this context a package feed is a "package source". I am just calling out that we remain consistent to call these sources and not feeds.

Reply:
@JonDouglas sorry, I wasn't clear enough in my question.

The restore summary today already says "NuGet feeds used".  Do you want me to change this to "NuGet package sources used"?  This can be done completely independently of this spec, since it's an existing message.

Since we're planning to filter the "audit sources list" to only sources with a VDB, I've convinced myself that we should no longer have a 1 line "package sources used as audit sources". The intent was to reduce pointless duplication when the two lists are the same (1 empty line, 1 heading line, 1 line per source). But since this is a signal of which sources contain a VDB now, I no longer think the 1 liner is a good idea.

If we kept this "Package sources used a audit sources" message , then I think it's important to use the same wording as what was shown a few lines above. It would not be clear to beginners if it says "NuGet feeds used: ... Package sources used as audit sources".  But since I'm no longer proposing this message, it's not very relevant to this spec any longer.

Reply:
Given the summary text proposal is to filter out sources that don't have a VDB, I'm withdrawing this "Package sources used for audit sources" message idea. If you, or anyone else, believes the existing "NuGet feeds used" message should be changed to "NuGet package sources used", please create a new issue.